Support integer globals, including integer comparisons

fleskesvor opened the pull request:
The goal of this PR is to properly support non-boolean globals so that it will eventually be possible to add counters (#66) and write more advanced comparisons in esc scripts.

This will require thorough testing, since there are many parts of the framework which rely on the return value from `vm.get_global()` being a global.

I have altered the autosave functionality to write the actual value of globals to the save script and also added comparisons using `eq [global] [value]` and `ne [global] [value]`. The latter does not perform any sanity checks yet, and should at the very least attempt to collect a quoted string containing comma(s).

I also haven't looked into regular save games yet, to make sure these changes don't cause any regressions.

fleskesvor commented on device/globals/esc_compile.gd:
```diff
@@ -256,21 +262,12 @@ func read_dialog_option(state, level, errors):
-		if if_inv.size():
-			cmd.if_inv = if_inv
-		if if_not_inv.size():
-			cmd.if_not_inv = if_inv
```
I think this must have been a bug, but since the `inv-inv_item` and `!inv-inv_item` is not documented anywhere, as far as I know, it was probably never encountered.

fleskesvor commented on device/globals/esc_compile.gd:
```diff
@@ -327,14 +321,9 @@ func read_cmd(state, level, errors):
-	if if_inv.size():
-		cmd.if_inv = if_inv
-	if if_not_inv.size():
-		cmd.if_not_inv = if_inv
```
Same presumed bug.

fleskesvor commented:
This PR should also fix a presumed bug where if one used the `!inv-inv_item` syntax instead of `!i/inv_item`, to check if `inv_item` is not in the inventory, the check for `inv-inv_item` would be run instead.

fleskesvor commented on device/globals/global_vm.gd:
```diff
@@ -306,7 +316,12 @@ func sched_event(time, obj, event):
+	if not name in globals:
+		return false
+	if typeof(globals[name]) == TYPE_BOOL:
+		return !!globals[name]
```
It's probably safer to treat all globals as strings going forward and compare the string to "true"+"false".

fleskesvor commented:
Added support for arithmetic operations in esc scripts using eg. `set_global my_counter +=1`, `set_global foo /=2`, etc. The second argument must match the regular expression `([-+*\/])=(\d+)`, so no spaces between `=` and the addend/multiplier/divisor/subtrahend.

mjtorn commented:
This looks pretty good!

One outstanding question: how does this discern between `1` and `"1"`?

I was thinking strings would need to have the double-quotes explicitly, so only integers and booleans would work without them, and trying to set something as `foo` would be an error.

Not sure we ever have to care about `"1"` anyway, but I couldn't find the code for integers so is that still underway?

Thanks anyway, good stuff :)

fleskesvor commented:
> One outstanding question: how does this discern between `1` and `"1"`?

It doesn't, all globals are strings, but cast to a boolean or number when convenient. I don't think this is an issue though, as long as we only use integers for numbers, since integers aren't subject to rounding errors, and equality checks will hold true whether they're actual integers or represented as strings. The number strings will only temporarily be cast to integers when context demand that they be numbers, like when using addition and subtraction, and doing number comparisons with lt, gt, le and ge.

fleskesvor commented:
This is pretty much done now. The only thing I can think of that might cause surprises is that comparison to a string containing a comma won't work, eg.:

`say player "Hello" [eq hello "Hello, World!"]`

I could fix this by extending the matched "flag" to include everything up until the second quotation mark, but I'm not sure if there are use cases for doing string comparisons anyway, and if so, should I also account for escaped quotation marks? Eg:

`say player "Hello" [eq hello "Hello, \"World\"!"]`

Anyway, I think this is ready for review now. I've tidied up my commits in a way that I think makes sense, but let me know if I should squash and/or rearrange anything.

mjtorn commented on device/globals/vm_level.gd:
```diff
@@ -29,7 +32,11 @@ func _walk(params, block):
 
 func set_global(params):
-	vm.set_global(params[0], params[1])
+	var result = _diff_pattern.search(params[1])
```
Is this maybe a bit generic with `_diff_pattern` and `result`?

```
var calc_operation = _calc_pattern.search(params[1])
if calc_operation:
    vm.set_global_calculate(params[0], calc_operation.get_string(1), calc_operation.get_string(2))
```

or somesuch variable names would be more obvious.

fleskesvor replied:
Yes, you're right. I will change the names to your suggestions.

mjtorn commented:
@fleskesvor I think beside the equality/comparison checking, the documentation should describe how to add or subtract.

Would you be open to a syntax like `incr_global i/pieces_of_eight 4` where the default to increment by would be `1` in the case of `incr_global i/pieces_of_eight` or just forcing it to always have an argument?

Not sure I got this right, but now it's `set_global i/pieces_of_eight +=4` without the space, right? That feels a bit different from the rest of ESC, and forbidding the space after the operator in ` += 4` feels a bit exotic in general.

Still loving the work, though!

fleskesvor commented:
> I think beside the equality/comparison checking, the documentation should describe how to add or subtract.

Yes, definitely.

> Would you be open to a syntax like `incr_global i/pieces_of_eight 4` where the default to increment by would be 1 in the case of `incr_global i/pieces_of_eight` or just forcing it to always have an argument?

Yes, I'm open to that, and I think it might actually be a better idea to use descriptive esc commands instead of adding "side effects" to `set_global`, even though it makes the language more verbose. Adding `incr_global` and `decr_global` (or similarly named commands) to the language would also remove the need for regular expressions or splitting strings to get at the value to add/subtract, so that's an added benefit. I think it would make sense to force the developer to use an integer as the second argument in that case though.

Thanks for the feedback, @mjtorn. I think I've also pretty much decided that putting in the work to improve string comparisons is not worth it for now, but maybe I'll write up what could be done in a feature request and leave that one for later.

mjtorn commented:
@fleskesvor sounds like a good plan from start to finish :) Hopefully moving to `inc_global` and `dec_global` isn't too big a hassle, and good thing to drop the r from the names!

Looking forward to testing this, there's a use case for me in my prototype game for counters :)

fleskesvor commented:
I've implemented the requested changes and documented the feature in `doc/esc_reference.md`.

fleskesvor commented on device/globals/esc_compile.gd:
```diff
@@ -205,15 +207,18 @@ func parse_flags(p_flags, flags_list, if_true, if_false, if_inv, if_not_inv):
 			else:
-				if_not_inv.push_back(trim(flag).substr(4, flag.length()-1))
+				ifs["if_not_inv"].push_back(trim(flag).substr(4, flag.length()-1))
+		if flag.substr(0, 3) in ["eq ", "ne ", "gt ", "ge ", "lt ", "le "]:
```
This should be an `elif`; otherwise negated conditions will be evaluated by the `else` block.

fleskesvor replied:
Another thing: Should it be possible to negate these new expressions? Possibly, and if so I should probably create a helper function for populating the lists.

fleskesvor commented:
I fixed the bug I introduced yesterday and added support for negating the new comparison operators. With that, having six separate operators became redundant, so I removed `ne`, `le` and `ge`. I'm now also trimming the expression after removing the negation operator (`!`), so it's now safe to use spaces between negation and expression, and that allowed me to reorder the order of the conditions within the block which parses negations to be consistent with the block for un-negated expressions.

This will require thorough testing though, since it touches on a core feature of Escoria. I have tested the new feature extensively, but have not tested properly for possible regressions yet, and would appreciate help with testing. The easiest way to test is to make an autosave with an item already in the inventory and one or two integer globals set, which can be decremented/incremented by interacting with an item in the demo scene. Testing should also include expressions with multiple conditions.

fleskesvor commented:
I have rebased on master, fixed the formatting on my additions to `doc/esc_reference.md` and squashed the commits I felt it made sense to squash. I have also done pretty extensive testing with combinations of boolean globals, inventory items and integer comparisons with variations with negations of all of the above, and it seems to be working as expected for me now. I believe this PR is ready to be merged now, and will do so before the weekend if there are no objections.

fleskesvor commented:
@mjtorn I made some of the commits from my work Mac, so it snuck in. :confused: I've removed it from the commit and added it to .gitignore in a separate commit. Merging this now that you have also tested it.
